### PR TITLE
Implement Band Matrix Multiply: All precisions & Tester

### PR DESCRIPTION
**Pull request by DSMishler**
Work on GBMM stretching back to ~July 2021. (repeat PR) - Major changes:

-     all gbmm functionality added (nonsquare, arbitrary kl,ku, all precisions, all transpose combinations)
-     gbmm tester added: also to all precisions. Format mimics that of gemm tester
-     relevant header files and python generating scripts modified to accommodate gbmm.

 

 Review would be most useful on:

 -     Double-checking that all supported build paths still work with gbmm (mainly CMake)
 -     Review on the high number of lines used on pzgemm
 -     Coding style. I would be happy to change any of the coding style to fit plasma style, but may not have always followed it. Force of habit.

  

  Commits that make this PR documented below in bullet point fashion

  -     Begin committing properly to git.
    -       Major changes to zgbmm.c. There may need to be other files created for clarity in its function
        -         transpose functionality erased from a run of codegen.py. I learned my lesson: I'll commit from now on.
          -       testing files: the generation of the band matrix isn't quite in order yet, but it is on the way
            -       header files: zgbmm.c and all other necessary test file adds added
            -     Documentation and investigation of testing code: still does not provide desired result.
            -     Test cases work: zlaset properly creates a band matrix and test code cleans up "corner bleeding" off initial set if needed. Band matrix multiply still relies heavily on general routines and may not be faster yet - that part is tested next.
            -     Implementing band matrix tiling: still don't quite have it figured out, but getting warmer. Version still works if matrix A is passed off as a general matrix.
            -     (DIRTY): progress and debug tools fill up the band matrix test environment, but issues have been identified. (This commit may be squashed later as it doesn't offer a deployable implementation of PLASMA)
            -     Lots of prints: working testing version with proper band tiling
            -     Removed uncommented print statements. May still have small bugs in need of investigation.
            -     Added support for nonsquare matrices - transposing support will be next
            -     Added transpose functionality; cleaned up debug/testing codes
            -     Added tuning (often unused) for gbmm. There is still the single precision gbmm bug remaining.
            -     fixed single precision passing error: simply added declarations to proper header files.

### CODE REVIEW

File: compute/zgbmm.c [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+    }
+    if (k < 0) {
+        plasma_error("illegal value of k");
+        return -5;
+    }
+    if (kl < 0) {
+        plasma_error("illegal value of kl");
+        return -6;
+    }
+    if (ku < 0) {
```
Comment (luszczek) [inline compute/zgbmm.c:156]: `ku` cannot exceed `m` or `n` depending on transpose parameters.

File: compute/zgbmm.c [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+    }
+    if (n < 0) {
+        plasma_error("illegal value of n");
+        return -4;
+    }
+    if (k < 0) {
+        plasma_error("illegal value of k");
+        return -5;
+    }
+    if (kl < 0) {
```
Comment (luszczek) [inline compute/zgbmm.c:152]: `kl` cannot exceed `m` or `n` depending on transpose parameters.

File: compute/zgbmm.c [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+    plasma_desc_destroy(&C);
+
+    // Return status.
+    int status = sequence.status;
+    return status;
+}
+
+/***************************************************************************//**
+ *
+ * @ingroup plasma_gemm
```
Comment (luszczek) [inline compute/zgbmm.c]: wrong group?

File: compute/zgbmm.c [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+ *          \f[ C = \alpha [op( A )\times op( B )] + \beta C, \f]
+ *
+ *  where op( X ) is one of:
+ *    \f[ op( X ) = X,   \f]
+ *    \f[ op( X ) = X^T, \f]
+ *    \f[ op( X ) = X^H, \f]
+ *
+ *  alpha and beta are scalars, and A, B and C are matrices, with op( A )
+ *  an m-by-k matrix, op( B ) a k-by-n matrix and C an m-by-n matrix.
+ *  A is a band Matrix, but the same is not guaranteed of B or C.
```
Comment (luszczek) [inline compute/zgbmm.c:39]: Why upper case Matrix?

File: compute/zgbmm.c [hunk 5]
```diff
@@ -0,0 +1,407 @@
...
+    }
+
+    // quick return
+    int k = transa == PlasmaNoTrans ? A.n : A.m;
+    if (C.m == 0 || C.n == 0 || ((alpha == 0.0 || k == 0) && beta == 1.0))
+        return;
+
+    // Call the parallel function.
+    // general band is part of this function: no need to change this one.
+    plasma_pzgemm(transa, transb,
```
Comment (luszczek) [inline compute/zgbmm.c:405]: wrong call?

File: test/test_zgbmm.c [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        (plasma_complex64_t*)malloc((size_t)ldc*Cn*sizeof(plasma_complex64_t));
+    assert(C != NULL);
+
+    int seed[] = {0, 0, 0, 1};
+    lapack_int retval;
+    /// need to determine what zlarnv does
+    retval = LAPACKE_zlarnv(1, seed, (size_t)lda*An, A);
+    assert(retval == 0);
+    /// The routine to set matrix values is laset (zlaset here)
+    LAPACKE_zlaset();
```
Comment (luszczek) [inline test/test_zgbmm.c]: empty call?

**Review by luszczek — COMMENTED** [review]
I started reviewing this but after going back and forth over 13 commits I'm not sure if some of the changes are supersede by other commits. This functionality is useful but I don't see very well what goes in and what goes out. I suggest to squash into GBMM, GBSET, and the tester for easier review.

**Comment by DSMishler** [comment]
Will be closing this PR and opening a simplified version in a few minutes. Would also like to discuss some of your comments shortly.

**Pull request closed without merging**